lib/uplink/main.go

```diff
@@ -0,0 +1,142 @@
+	// files should come from the metainfo index. It's unlikely these will ever
+	// change much.
+	EncBlockSize  int
+	MaxInlineSize int
+	SegmentSize   int64
+	TLSConfig     tlsopts.Config
+}
+
+// Session represents a specific access session.
+type Session struct {
+	TransportClient transport.Client
+	Gateway         *miniogw.Gateway
+}
+
+// Access is all of the access information an application needs to store and
+// retrieve data. Someone with a share may have no restrictions within a project
+// (can create buckets, list buckets, list files, upload files, delete files,
+// etc), may be restricted to a single bucket, may be restricted to a prefix
+// within a bucket, or may even be restricted to a single file within a bucket.
+// NB(dylan): You need an Access to start a Session
+type Access struct {
+	Permissions Macaroon
+
+	// TODO: these should be per-bucket somehow maybe? oh man what a nightmare
+	// Could be done via []Bucket with struct that has each of these
+	// PathCipher       storj.Cipher // i.e. storj.AESGCM
+	// EncPathPrefix    storj.Path
+	// Key              storj.Key
+	// EncryptionScheme storj.EncryptionScheme
+
+	// Something like this?
+	// TODO(dylan): Shouldn't actually use string, this is just a placeholder
+	// until a more precise type is figured out - probably type Bucket
+	Buckets map[string]BucketOpts
+}
+
+// Uplink represents the main entrypoint to Storj V3. An Uplink connects to
+// a specific Satellite and caches connections and resources, allowing one to
+// create sessions delineated by specific access controls.
+type Uplink struct {
+	ID            *identity.FullIdentity
+	Session       *Session
+	SatelliteAddr string
+	Config        Config
+}
+
+// NewUplink creates a new Uplink
+func NewUplink(ident *identity.FullIdentity, satelliteAddr string, cfg Config) *Uplink {
+	return &Uplink{
+		ID:            ident,
+		SatelliteAddr: satelliteAddr,
+		Config:        cfg,
+	}
+}
+
+// ParseAccess parses a serialized Access
+func ParseAccess(data []byte) (Access, error) {
+	panic("TODO")
+}
+
+// Serialize serializes an Access message
+func (a *Access) Serialize() ([]byte, error) {
+	panic("TODO")
+}
+
+// NewSession creates a Session with an Access struct.
+func (u *Uplink) NewSession(ctx context.Context, access Access) error {
+	opts, err := tlsopts.NewOptions(u.ID, u.Config.TLSConfig)
+	if err != nil {
+		return err
+	}
+
+	tc := transport.NewClient(opts)
+
+	gateway, err := getGateway(ctx, u.ID)
+	if err != nil {
+		return err
+	}
+
+	u.Session = &Session{
+		TransportClient: tc,
+		Gateway:         gateway,
+	}
+
+	return nil
+}
+
+// Access creates a new share, potentially further restricted from the Access used
+// to create this session.
+func (s *Session) Access(ctx context.Context, caveats ...Caveat) (Access, error) {
+	panic("TODO")
+}
+
+func getGateway(ctx context.Context, identity *identity.FullIdentity) (*miniogw.Gateway, error) {
```

Comment:
this package should have no dependency on miniogw at all. instead, the code in miniogw should be moved to this package and miniogw should use this package

Reply:
Ah, I think I might have misunderstood how this was supposed to fit into our ecosystem then. I can work on restructuring this then. Back to the drawing board.